server/src/com/thoughtworks/go/server/materials/ConfigMaterialUpdater.java

```diff
@@ -0,0 +1,135 @@
+                                 MaterialChecker materialChecker,
+                                 ConfigMaterialUpdateCompletedTopic configCompletedTopic,
+                                 MaterialUpdateCompletedTopic topic,
+                                 MaterialService materialService,
+                                 SubprocessExecutionContext subprocessExecutionContext)
+    {
+        this.repoConfigDataSource = repoConfigDataSource;
+        this.materialChecker = materialChecker;
+        this.materialRepository = materialRepository;
+        this.configCompleted = configCompletedTopic;
+        this.topic = topic;
+        this.materialService = materialService;
+        this.subprocessExecutionContext = subprocessExecutionContext;
+
+        this.configCompleted.addListener(this);
+    }
+
+    @Override
+    public void onMessage(MaterialUpdateCompletedMessage message) {
+        Material material = message.getMaterial();
+        //MDU is done using the checkout, it has done db update and stored latest changes
+        // but MUS is still waiting for material updated message on MaterialUpdateCompletedTopic
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(format("[Config Material Update] Config material update completed for material %s. Starting parse process", material));
+        }
+        try {
+            if(message instanceof MaterialUpdateFailedMessage)
+            {
+                MaterialUpdateFailedMessage failure = (MaterialUpdateFailedMessage)message;
+                LOGGER.warn(String.format("[Config Material Update] Cannot update configuration part because material update has failed. Reason: ",
+                        failure.getReason()));
+            }
+            else {
+                File folder = materialRepository.folderFor(material);
+                MaterialRevisions latestModification = materialRepository.findLatestModification(material);
+                Revision revision = latestModification.firstModifiedMaterialRevision().getRevision();
+
+                MaterialRevision lastParseRevision = getMaterialRevisionAtLastParseAttempt(message);
+                if (lastParseRevision == null) {
+                    //never parsed
+                    UpdateConfigurationFromCheckout(folder, revision, material);
+                } else if (latestModification.findRevisionFor(material.config())
+                        .hasChangedSince(lastParseRevision)) {
+                    // revision has changed. the config files might have been updated
+                    UpdateConfigurationFromCheckout(folder, revision, material);
+                } else {
+                    // revision is the same as last time, no need to parse again
+                }
+            }
+        }
+        finally {
+            // always post the original message further
+            // this will remove material from inProgress in MUS
+            topic.post(message);
+        }
+    }
+
```

Comment:
So, `onMessage(MaterialUpdateCompletedMessage message)` would happen on a successful/failed MDU of a config-repo-material. In case of a successful MDU, the code seems to be calling `UpdateConfigurationFromCheckout ` which calls `poller.checkout` all over again. The folder here is the flyweight folder which would have been updated during the MDU already. Is that intentional? I am still trying to get my head around this, so there is a chance I might have missed out on one of your comments.